xapian-bindings/java/java.i

```diff
@@ -174,21 +174,63 @@ namespace Xapian {
     bool hasNext() const { return Xapian::iterator_valid(*self); }
 }
 
+%ignore MSetIterator::next();
+%javamethodmodifiers MSetIterator::next_helper() "private"
+
 %extend MSetIterator {
-    Xapian::docid next () {
-	Xapian::docid tmp;
+    void next_helper () {
 	if (Xapian::iterator_valid(*self)) {
-	    tmp = (**self);
 	    ++(*self);
-	} else {
-	    tmp = -1;
 	}
-	return tmp;
     }
 
     bool hasNext() const { return Xapian::iterator_valid(*self); }
 }
 
+%typemap(javaimports) Xapian::MSetIterator "import java.util.Iterator;"
+%typemap(javainterfaces) Xapian::MSetIterator "Iterator<MSetIterator.MatchDescriptor>"
+%typemap(javacode) MSetIterator%{
+    @Override
+    public MatchDescriptor next() {
+        MatchDescriptor descriptor = new MatchDescriptor();
+        if(hasNext()) {
```

Comment:
I feel this check may not be necessary because we are already checking the iterator validity within the next_helper function before moving to the next item.